### PR TITLE
Bug fixes in General Summary Dialogue

### LINKED ISSUE #8744: Prepare > Data Reshape > General Summaries needs bugs fixing
This is an incredibly powerful dialog.  Once fixed it will still be a pain to use, but we could use it to calculate the complicated summaries that we use, for example, for the start of the rains.  So it really shows the power and potential of R-Instat.  I found the bugs when trying to write the help for the dialog.  I mention 2 (of the?) bugs below.  I think if those are fixed it should be easy to proceed and for the team to do anything more - if more is needed.
@ChrisMarsh82 David has (I am afraid) proposed you to fix the obvious bugs, given below.  
@volloholic is hoping you could have a quick look and either fix them - he thinks that should be easy for you.  Or you then allocate.

The example shown here is trivial, but demonstrates that the system does work.  It uses the ggplot2 diamonds data and the caclulator and gives the mean price for each level of the color factor. 

Here is the output it produced:
image

And here is the (corrected) script, which produced it.

```
# Dialog: General Summaries

calc1 <- instat_calculation$new(name="calc1", type="summary", save=2, result_name="bob", manipulations=list(manip1=instat_calculation$new(name="manip1", type="by", save=0, result_name="", calculated_from=list(diamonds = "color"))), function_exp="mean(price , na.rm = TRUE)")
data_book$run_instat_calculation(calc=calc1, display=TRUE)

rm(calc1)
```
Now this was corrected by David.  The script directly from the dialog was as follows:

```
# Dialog: General Summaries

calc1 <- instat_calculation$new(name="calc1", type="summary", save=1, result_name="", manipulations=list(manip1=instat_calculation$new(name="manip1", type="by", save=0, result_name="", calculated_from=list(diamonds = "color",diamonds = ""))), function_exp="mean(price , na.rm = TRUE)")
data_book$run_instat_calculation(calc=calc1, display=TRUE)

rm(calc1)
```
So, he made 2 changes.  The first was to add a result name on the first line.  The second was to delete the `,diamonds-""` on line 3 

Now the 2 or 3 bugs so far.  Go into the dialog and click on Add.  You then get this:
image

It doesn't find the diamonds data.  That's when the type is calculation or summary.
If you change to By, then it does find the data and gives the 3 factor variables.

Now return to the calculation, where you can go to any keyboard.  Go to the summary keyboard and press the mean key.

Now you get an error that probably stops R-Instat on the merged version:

image

Of course that's an invalid calculation, because it doesn't have anything to take the mean of.  But it shouldn't be trying to execute anything at this point anyway.  You can actually type price (which is a variable name) there and you get the same nasty message each letter you type.  It should just be sitting there patiently!

I think there is more to correct, but if we could fix these, then we could, at least, test the dialogue.

### PR DESCRIPTION
**Pull request by N-thony**
Fixes #8744
@rdstern I don't get the script you have posted in the issue but I fixed some bugs reported there. Have a look

### CODE REVIEW

**Comment by rdstern** [comment]
@N-thony there is one other obvious problem to fix - if you can, before I discuss with David?

image

If I load the diamonds, go into the dialog and press Add I get this, and without the dataset.  I assume it should access the diamonds data in the control, as the data selector usually does.  I wonder why not?

**Review by rdstern — COMMENTED** [review]
@N-thony excellent.  I still don't know how to use it, or if the results are correct.  I'll discuss with @volloholic and then comment further.

**Comment by rdstern** [comment]
@N-thony thanks for the improvements so far.  I have been through a calculation with David, and can now get it to work, for something simple, by making the corrections in the script window.  The current problems may have been there all along, or perhaps your corrections may have had an unfortunate side effect.  I take you through this calculation step-by-step in case you can see obvious bugs in the front-end, once you see the errors in the generated code.

a) So diamonds data - always start on that data frame
b) Dialog > Add > Manipulations > as shown below:  (any number of the factors.  I have tried with just one, this is my first time with all 3.)

image

c) Return and back to Main and Type = Summary:

image

d) Get the mean of price in the calculator - call the result antoine as shown:

image

e) Return and To Script:

Here is the script I get:

```
# Dialog: General Summaries

calc1 <- instat_calculation$new(name="calc1", type="summary", save=2, result_name="antoine", manipulations=list(manip1=instat_calculation$new(name="manip1", type="by", save=0, result_name="", calculated_from=list(diamonds = "color",diamonds = "clarity",diamonds = "cut",))), function_exp="mean(price , na.rm = TRUE)", calculated_from=)
data_book$run_instat_calculation(calc=calc1, display=FALSE)
rm(calc1)
```
There are 2 problems to be corrected in the script:
a) There is an extra comma at the end here:  `calculated_from=list(diamonds = "color",diamonds = "clarity",diamonds = "cut",`
b) This is blank  `calculated_from=` at the end of the command.  It should (I think) be: `calculated_from=list(diamonds = "color",diamonds = "clarity",diamonds = "cut",diamonds = "price")`  
(I notice that just  `calculated_from=list(diamonds = "price")`  will also run, but David thinks that the general system may sometimes need the full list.)

Then I get the result:

image

We are hoping that you can see some obvious places in the code where these bugs arise?  Good luck!  I am also happy to take you through these steps on my screen if that would help.  I would very much like to get this powerful system working so it can also be documented.  It is quite a challenge without Danny.

**Review by rdstern — COMMENTED** [review]
@N-thony that's brilliant and a great relief.  I am discussing again with @volloholic this afternoon and will report in more detail then.
But so good that it is starting to work again.

One "small?" thing you may be able to improve quickly is as follows.  Here is the code for something that works fine now:

```
# Dialog: General Summaries
calc1 <- instat_calculation$new(name="calc1", type="summary", save=2, result_name="gg", calculated_from=list(diamonds = "carat",diamonds = "cut",diamonds = "color",diamonds = "clarity",diamonds = "depth",diamonds = "table",diamonds = "price",diamonds = "x",diamonds = "y",diamonds = "z",diamonds = "james"), manipulations=list(manip1=instat_calculation$new(name="manip1", type="by", save=0, result_name="", calculated_from=list(diamonds_by_cut_color_clarity = "cut",diamonds_by_cut_color_clarity = "color",diamonds_by_cut_color_clarity = "clarity"))), function_exp="min(depth , na.rm = TRUE)")
data_book$run_instat_calculation(calc=calc1, display=FALSE)
```
But I notice the first `calculated from` includes all the variables in the dataframe.  So I tried again with the efc data from sjplot package that has 26 variables and it is here:

```
# Dialog: General Summaries
calc1 <- instat_calculation$new(name="calc1", type="summary", save=2, result_name="roger", calculated_from=list(efc = "c12hour",efc = "e15relat",efc = "e16sex",efc = "e17age",efc = "e42dep",efc = "c82cop1",efc = "c83cop2",efc = "c84cop3",efc = "c85cop4",efc = "c86cop5",efc = "c87cop6",efc = "c88cop7",efc = "c89cop8",efc = "c90cop9",efc = "c160age",efc = "c161sex",efc = "c172code",efc = "c175empl",efc = "barthtot",efc = "neg_c_7",efc = "pos_v_4",efc = "quol_5",efc = "resttotn",efc = "tot_sc_e",efc = "n4pstu",efc = "nur_pst"), manipulations=list(manip1=instat_calculation$new(name="manip1", type="by", save=0, result_name="", calculated_from=list(efc = "e15relat",efc = "e16sex"))), function_exp="mean(e17age , na.rm = TRUE)")
data_book$run_instat_calculation(calc=calc1, display=FALSE)
```
 
So clearly it will be unmanageable for any wide dataset.  Could you change this first?

**Comment by rdstern** [comment]
@N-thony yes, just make the list of variables include those that are needed in that operation?  Remember the data frame may have even 1000's of variables!

**Review by rdstern — DISMISSED** [review]
@N-thony great, that looks neat and still gets the answers for the simple example I am using.  @Patowhiz could you do a quick check and then merge?  The current code and dialog don't work at all, so this has to be better!

File: instat/sdgCalculationsSummmary.vb [hunk 1]
```diff
@@ -61,7 +61,7 @@ Public Class sdgCalculationsSummmary
 
         ucrManipulations.lstAvailableData.View = View.List
 
-        ucrCalcSummary.ucrReceiverForCalculation.bAttachedToPrimaryDataFrame = False
+        'ucrCalcSummary.ucrReceiverForCalculation.bAttachedToPrimaryDataFrame = False
```
Comment (Patowhiz) [inline instat/sdgCalculationsSummmary.vb:64]: Remove this entirely if comment is not for future reference?

File: instat/sdgCalculationsSummmary.vb [hunk 2]
```diff
@@ -374,14 +376,23 @@ Public Class sdgCalculationsSummmary
...
+            Return ""
+        End If
+
         Dim strCalcFromList As String
 
         strCalcFromList = "list("
         For i = 0 To lstVariables.Count - 1
-            If i > 0 Then
-                strCalcFromList = strCalcFromList & ","
+            Dim strDataName = lstVariables(i)
```
Comment (Patowhiz) [inline instat/sdgCalculationsSummmary.vb:388]: Explicitly define the variable type for self documenting code.

**Review by Patowhiz — CHANGES_REQUESTED** [review]
@N-thony looks good just a few comments and suggestions.

**Review by rdstern — APPROVED** [review]
@Patowhiz still seems ok to me, so approving again

**Pull request merged**